dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java

```diff
@@ -0,0 +1,222 @@
+import org.eclipse.jetty.http2.client.HTTP2Client;
+import org.eclipse.jetty.http2.frames.DataFrame;
+import org.eclipse.jetty.http2.frames.HeadersFrame;
+import org.eclipse.jetty.http2.frames.ResetFrame;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.FuturePromise;
+import org.eclipse.jetty.util.ssl.SslContextFactory;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class DicomWebClientJetty implements IDicomWebClient {
+
+  private static final int CONNECT_PORT = 443;
+
+  private final String serviceUrlPrefix;
+  private final OAuth2Credentials credentials;
+
+  public DicomWebClientJetty(
+      OAuth2Credentials credentials,
+      String serviceUrlPrefix) {
+    this.credentials = credentials;
+    this.serviceUrlPrefix = Util.trim(serviceUrlPrefix);
+  }
+
+  @Override
+  public MultipartInput wadoRs(String path) throws DicomWebException {
+    throw new UnsupportedOperationException("Not Implemented, use DicomWebClient");
+  }
+
+  @Override
+  public JSONArray qidoRs(String path) throws DicomWebException {
+    throw new UnsupportedOperationException("Not Implemented, use DicomWebClient");
+  }
+
+  @Override
+  public void stowRs(String path, InputStream in) throws DicomWebException {
+    try {
+      HTTP2Client client = new HTTP2Client();
+      SslContextFactory sslContextFactory = new SslContextFactory.Client();
+      client.addBean(sslContextFactory);
+      client.start();
+
+      HttpURI uri = new HttpURI(serviceUrlPrefix + "/" + Util.trim(path));
+
+      FuturePromise<Session> sessionPromise = new FuturePromise<>();
+      client.connect(sslContextFactory, new InetSocketAddress(uri.getHost(), CONNECT_PORT),
+          new ServerSessionListener.Adapter(), sessionPromise);
+      Session session = sessionPromise.get(5, TimeUnit.SECONDS);
+
+      // Prepare the request
+      HttpFields requestFields = new HttpFields();
+      if (credentials != null) {
+        credentials.getRequestMetadata();
+        requestFields.add(HttpHeader.AUTHORIZATION,
+            "Bearer " + credentials.getAccessToken().getTokenValue());
+      }
+      requestFields.add(HttpHeader.CONTENT_TYPE,
+          "application/dicom");
+      requestFields.add(HttpHeader.ACCEPT,
+          "application/dicom+json");
+      MetaData.Request request = new MetaData.Request("POST", uri, HttpVersion.HTTP_2,
+          requestFields);
+      HeadersFrame headersFrame = new HeadersFrame(request, null, false);
+
+      // Prepare the listener to receive the HTTP response frames.
+      final StringBuilder resultBuilder = new StringBuilder();
+      final CompletableFuture<Integer> responseCodeFuture = new CompletableFuture<>();
+      final CompletableFuture<Boolean> doneFuture = new CompletableFuture<>();
+      Stream.Listener responseListener = new Stream.Listener.Adapter() {
+        @Override
+        public void onReset(Stream stream, ResetFrame frame) {
```

Comment:
When does onReset get called?

Reply:
When error response comes during post body transmission. Frame sequence for error is Header  -> Data (not end stream)  -> Reset.
For normal response it's Header  -> Data (end stream).